File: mamda/dotnet/src/NUnittest/secstatus/MamdaSecStatusListenerShortSaleTests.cs
using System;
using System.Collections.Generic;
using System.Collections;
using NUnit.Framework;
using Wombat;

namespace NUnitTest
{
	[TestFixture]
    public class MamdaSecStatusListenerShortSaleTests
    {
        MamaMsg mMsg = null;
        MamdaSubscription mSubscription = null;
        MamdaSecurityStatusListener mSecStatusListener = null;
          
        [SetUp]
        public void SetUp()
        {
            try
            {
                MamaBridge mBridge = Mama.loadBridge("lbm");
                Mama.open();

				MamaDictionary mDictionary = new MamaDictionary();
                mDictionary.create("dictionary.txt");

				MamdaCommonFields.setDictionary(mDictionary, null);
                MamdaSecurityStatusFields.reset();
                MamdaSecurityStatusFields.setDictionary(mDictionary, null);

                mSubscription = new MamdaSubscription();                             
            }
            catch (Exception ex)
            {
                throw new MamdaDataException(ex.Message, ex);
            }
        }

        public void callMamdaOnMsg(MamdaSubscription sub, MamaMsg msg)
        {
            try
            {
                ArrayList msgListeners = new ArrayList();
                msgListeners = sub.getMsgListeners();
                IEnumerator iter = msgListeners.GetEnumerator();
                int size = msgListeners.Count;
                for (int i = 0; i < size; i++)
                {
                    iter.MoveNext();
                    MamdaMsgListener listener = (MamdaMsgListener)iter.Current;
                    listener.onMsg(sub, msg, msg.getType());
                }
            }
            catch (Exception ex)
            {
                throw new MamdaDataException(ex.Message, ex);
            }
        }

        public void addMamaHeaderFields(
                            MamaMsg msg,
                            byte msgType,
                            byte msgStatus,
                            uint seqNum)
        {
            msg.addU8(null, 1, msgType);
            msg.addU8(null, 2, msgStatus);
            msg.addU32(null, 10, seqNum);
            msg.addU64(null, 20, 1);
        }

        public void clear()
        {
            mMsg.clear();
            addMamaHeaderFields(mMsg,
                                   1,
                                   0,
                                   1);
        }

        [Test]
        public void testNotInEffect()
        {
            mSecStatusListener = new MamdaSecurityStatusListener();
            mSubscription.addMsgListener(mSecStatusListener);            
            mMsg = new MamaMsg();
            addMamaHeaderFields(mMsg,
                                (byte)mamaMsgType.MAMA_MSG_TYPE_SEC_STATUS,
                                (byte)mamaMsgStatus.MAMA_MSG_STATUS_OK,
                                (uint)1);               
            callMamdaOnMsg(mSubscription, mMsg);
            Assert.AreEqual(' ', mSecStatusListener.getShortSaleCircuitBreaker());
            Assert.AreEqual(MamdaFieldState.NOT_INITIALISED, 
                                mSecStatusListener.getShortSaleCircuitBreakerFieldState());
            mMsg = null;
            mSecStatusListener = null;
        }

        [Test]
        public void testCached()
        {
            mSecStatusListener = new MamdaSecurityStatusListener();
            mSubscription.addMsgListener(mSecStatusListener);          
              
            mMsg = new MamaMsg();
            addMamaHeaderFields(mMsg,
                                (byte)mamaMsgType.MAMA_MSG_TYPE_SEC_STATUS,
                                (byte)mamaMsgStatus.MAMA_MSG_STATUS_OK,
                                (uint)1);
            mMsg.addChar(null, 5248, 'A');
            callMamdaOnMsg(mSubscription, mMsg);
            Assert.AreEqual('A', mSecStatusListener.getShortSaleCircuitBreaker());
            Assert.AreEqual(MamdaFieldState.MODIFIED,
                                mSecStatusListener.getShortSaleCircuitBreakerFieldState());
            mMsg = null;
            mSecStatusListener = null;
        }

        [Test]
        public void testUpdated()
        {
            mSecStatusListener = new MamdaSecurityStatusListener();
            mSubscription.addMsgListener(mSecStatusListener);            
                
            mMsg = new MamaMsg();
            addMamaHeaderFields(mMsg,
                                (byte)mamaMsgType.MAMA_MSG_TYPE_SEC_STATUS,
                                (byte)mamaMsgStatus.MAMA_MSG_STATUS_OK,
                                (uint)1);
            mMsg.addChar(null, 5248, 'C');
            callMamdaOnMsg(mSubscription, mMsg);
            Assert.AreEqual('C', mSecStatusListener.getShortSaleCircuitBreaker());
            Assert.AreEqual(MamdaFieldState.MODIFIED,
                                mSecStatusListener.getShortSaleCircuitBreakerFieldState());

            mMsg = null;
            mMsg = new MamaMsg();
            addMamaHeaderFields(mMsg,
                                (byte)mamaMsgType.MAMA_MSG_TYPE_SEC_STATUS,
                                (byte)mamaMsgStatus.MAMA_MSG_STATUS_OK,
                                (uint)2);
            callMamdaOnMsg(mSubscription, mMsg);
            Assert.AreEqual(MamdaFieldState.NOT_MODIFIED, 
                            mSecStatusListener.getShortSaleCircuitBreakerFieldState());
            mMsg = null;

            mMsg = new MamaMsg();
            addMamaHeaderFields(mMsg,
                                (byte)mamaMsgType.MAMA_MSG_TYPE_SEC_STATUS,
                                (byte)mamaMsgStatus.MAMA_MSG_STATUS_OK,
                                (uint)3);
            mMsg.addChar(null, 5248, 'D');
            callMamdaOnMsg(mSubscription, mMsg);
            Assert.AreEqual('D', mSecStatusListener.getShortSaleCircuitBreaker());
            Assert.AreEqual(MamdaFieldState.MODIFIED,
                                mSecStatusListener.getShortSaleCircuitBreakerFieldState());
            mMsg = null;
            mSecStatusListener = null;
        }     
    }
}

File: mamda/c_cpp/src/gunittest/secstatus/MamdaSecStatusListenerShortSaleTests.cpp
/* $Id$
 *
 * OpenMAMA: The open middleware agnostic messaging API
 * Copyright (C) 2011 NYSE Technologies, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 */

#include <gtest/gtest.h>

#include <mamda/MamdaSecStatusFields.h>
#include <mamda/MamdaCommonFields.h>
#include <mamda/MamdaSubscription.h>
#include <mamda/MamdaConfig.h>
#include <mamda/MamdaSecStatusListener.h>
#include <mamda/MamdaSecStatusHandler.h>
#include <mamda/MamdaDataException.h>
#include <mamda/MamdaFieldState.h>
#include <mamda/MamdaErrorListener.h>
#include <mamda/MamdaQualityListener.h>
#include <mama/MamaDictionary.h>
#include <mama/mamacpp.h>
#include <mama/MamaMsg.h>

#include "common/MamdaUnitTestUtils.h"
#include "common/CpuTestGenerator.h"
#include "common/MemoryTestGenerator.h"

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <cstdlib>
#include <sstream>
#include <iostream>
#include <fstream>

using std::string;

using namespace std;
using namespace Wombat;

class MamdaSecStatusListenerShortSaleTests : public ::testing::Test
{
protected:
    MamdaSecStatusListenerShortSaleTests () {}
    virtual ~MamdaSecStatusListenerShortSaleTests () {}  

    virtual void SetUp()
    {
        try
        {
            mamaBridge bridge;
            bridge = Mama::loadBridge("wmw");
            Mama::open();
            //mama_enableLogging (stderr, MAMA_LOG_LEVEL_FINEST);
            mDictionary = new MamaDictionary;
            if (!mDictionary)
            {
                FAIL() << "Failed to allocate MamaDictionary\n";
                return;
            }
            
            mDictionary->populateFromFile("dictionary.txt");
            MamdaCommonFields::setDictionary (*mDictionary);
            MamdaSecStatusFields::reset();
            MamdaSecStatusFields::setDictionary (*mDictionary);
                      
            mSubscription = new MamdaSubscription;
            if (!mSubscription)
            {
                FAIL() << "Failed to allocate MamdaSubscription\n";
                return;
            }
            mSecStatusListener = new MamdaSecStatusListener; 
            if (!mSecStatusListener)
            {
                FAIL() << "Failed to allocate MamdaTradeListener\n";
                return;
            }
            mSubscription->addMsgListener(mSecStatusListener);
          
        }
        catch(MamaStatus status)
        {
            FAIL() << "Failed to setup Mamda: "
                   << status.toString() << "\n";
            return;
        }
    }    
    
    virtual void TearDown()
    {     
        if (mDictionary)
        {     
            delete mDictionary;
            mDictionary = NULL;
        }  
        if (mSecStatusListener)
        {
            delete mSecStatusListener;
            mSecStatusListener = NULL;
        }        
    }        

    MamaDictionary*         mDictionary;
    MamdaSecStatusListener* mSecStatusListener;
    MamdaSubscription*      mSubscription;
};

void callMamdaOnMsg (MamdaSubscription* sub, MamaMsg& msg)
{  
    try
    {   
        vector<MamdaMsgListener*>& msgListeners = sub->getMsgListeners();
        unsigned long size = msgListeners.size();
        for (unsigned long i = 0; i < size; i++)           
        {        
            MamdaMsgListener* listener = msgListeners[i];
            listener->onMsg (sub, msg,(short)msg.getType());
        }   
    }                
    catch (std::exception &e)
    {
        std::cerr << "Exception: " << e.what() << std::endl;
    }
    catch (...)
    {
        std::cerr << "Exception: Unknown" << std::endl;
    }
    return;
}      


TEST_F(MamdaSecStatusListenerShortSaleTests, TestShortSaleCircuitBreaker)
{
    MamaMsg* mMsg = NULL;
    mMsg = new MamaMsg;
    mMsg->create();
    addMamaHeaderFields(*mMsg,
                        MAMA_MSG_TYPE_SEC_STATUS,
                        MAMA_MSG_STATUS_OK,
                        1);
                        
    callMamdaOnMsg(mSubscription,*mMsg);  
    EXPECT_TRUE(' ' == mSecStatusListener->getShortSaleCircuitBreaker());
    EXPECT_EQ(NOT_INITIALISED,mSecStatusListener->getShortSaleCircuitBreakerFieldState());   
                        
    mMsg->addChar("wShortSaleCircuitBreaker",5248,'A');    
    callMamdaOnMsg(mSubscription,*mMsg);  
    EXPECT_TRUE('A' == mSecStatusListener->getShortSaleCircuitBreaker());
    EXPECT_EQ(MODIFIED,mSecStatusListener->getShortSaleCircuitBreakerFieldState());          
    
    mMsg->updateChar("wShortSaleCircuitBreaker",5248,'D');    
    callMamdaOnMsg(mSubscription,*mMsg);  
    EXPECT_TRUE('D' == mSecStatusListener->getShortSaleCircuitBreaker());
    EXPECT_EQ(MODIFIED,mSecStatusListener->getShortSaleCircuitBreakerFieldState());  
}



